### PR TITLE
ivy-component-metadata-rule sample should check for IvyModuleDescriptor existence to avoid breaking Maven modules

### PR DESCRIPTION
**Pull request by rpalcolea**
This enhance docs around `ivy-component-metadata-rule` to prevent breaking Maven based modules

If check for `IvyModuleDescriptor` is not in place, the rule will fail with maven based modules causing issues such as:

```
Execution failed for task ':compileJava'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve com.google.guava:guava:19.0.
     Required by:
         project : > foo:my-core:1.0.0
      > Cannot choose between the following variants of com.google.guava:guava:19.0:
          - apiElements
          - compile
          - runtime
        All of them match the consumer attributes:
          - Variant 'apiElements' capability com.google.guava:guava:19.0:
              - Unmatched attributes:
                  - Required org.gradle.dependency.bundling 'external' but no value provided.
``` 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/docs/src/snippets/userguide/dependencyManagement/customizingResolution/ivyMetadataRule/groovy/build.gradle [hunk 1]
```diff
@@ -15,6 +15,10 @@ class IvyVariantDerivationRule implements ComponentMetadataRule {
     @Inject ObjectFactory getObjects() { }
 
     void execute(ComponentMetadataContext context) {
+        if(context.getDescriptor(IvyModuleDescriptor) == null) {
+            return
```
Comment (ljacomet) [inline subprojects/docs/src/snippets/userguide/dependencyManagement/customizingResolution/ivyMetadataRule/groovy/build.gradle]: I would add a comment that indicates this tests filters out any non Ivy module.

**Comment by ljacomet** [comment]
And in addition, please target `master` for this instead of `release`.

**Comment by rpalcolea** [comment]
> And in addition, please target `master` for this instead of `release`.

Hey @ljacomet , added the comments and changed it to target `master`.

File: subprojects/docs/src/snippets/userguide/dependencyManagement/customizingResolution/ivyMetadataRule/kotlin/build.gradle.kts [hunk 2]
```diff
@@ -15,6 +15,11 @@ open class IvyVariantDerivationRule : ComponentMetadataRule {
     @Inject open fun getObjects(): ObjectFactory = throw UnsupportedOperationException()
 
     override fun execute(context: ComponentMetadataContext) {
+        // This filters out any non Ivy module
+        if(context.getDescriptor(IvyModuleDescriptor) == null) {
```
Comment (ljacomet) [inline subprojects/docs/src/snippets/userguide/dependencyManagement/customizingResolution/ivyMetadataRule/kotlin/build.gradle.kts]: This is not valid in Kotlin. You have to use `context.getDescriptor(IvyModuleDescriptor::class)`

Comment (rpalcolea) [reply subprojects/docs/src/snippets/userguide/dependencyManagement/customizingResolution/ivyMetadataRule/kotlin/build.gradle.kts]: oops! nice catch. fixed in b1b94ef

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@rpalcolea I have rebased your PR on top of master as you were hit by unrelated errors. Assuming the build passes, I will merge.

**Comment by rpalcolea** [comment]
thanks for looking at this @ljacomet  👍 

Let me know if you get a chance to look at https://github.com/gradle/gradle/pull/12480 😅

**Pull request merged**